src/D2L.Bmx/UpdateHandler.cs

```diff
@@ -0,0 +1,129 @@
+using System.Formats.Tar;
+using System.IO.Compression;
+using System.Reflection;
+using System.Runtime.InteropServices;
+
+namespace D2L.Bmx;
+
+internal class UpdateHandler {
+
+	public async Task HandleAsync() {
+		using var httpClient = new HttpClient();
+		GithubRelease? releaseData = await GithubRelease.GetLatestReleaseDataAsync();
+		Version? latestVersion = releaseData?.GetReleaseVersion();
+		if( latestVersion is null ) {
+			throw new BmxException( "Failed to grab the latest version" );
+		}
+
+		var localVersion = Assembly.GetExecutingAssembly().GetName().Version;
+		if( latestVersion <= localVersion ) {
+			Console.WriteLine( $"Already own the latest version {latestVersion}" );
+			return;
+		}
+
+		string archiveName = GetOSFileName();
+		string? downloadUrl = releaseData?.Assets?.FirstOrDefault( a => a.Name == archiveName )?.BrowserDownloadUrl;
+		if( string.IsNullOrWhiteSpace( downloadUrl ) ) {
+			throw new BmxException( "Failed to get update download url" );
+		}
+
+		string? currentFilePath = Environment.ProcessPath;
+		if( string.IsNullOrEmpty( currentFilePath ) ) {
+			throw new BmxException( "BMX could not update" );
+		}
+
+		string currentDirectory = Path.GetDirectoryName( currentFilePath )!;
+		long backupPathTimeStamp = DateTime.Now.Millisecond;
+		string backupPath = Path.Join( BmxPaths.OLD_BMX_VERSIONS_PATH, $"bmx-v{localVersion}-{backupPathTimeStamp}-old.bak" );
+		try {
+			File.Move( currentFilePath, backupPath );
+		} catch( IOException ex ) {
+			throw new BmxException( "Could move the old version, try with elevated permissions", ex );
+		} catch {
+			throw new BmxException( "BMX could not update" );
+		}
+
+
+		string downloadPath = Path.GetTempFileName();
+		try {
+			var archiveRes = await httpClient.GetAsync( downloadUrl, HttpCompletionOption.ResponseHeadersRead );
+			using var fs = new FileStream( downloadPath, FileMode.Create, FileAccess.Write, FileShare.None );
+			await archiveRes.Content.CopyToAsync( fs );
+			await fs.FlushAsync();
+		} catch( Exception ex ) {
+			File.Move( backupPath, currentFilePath );
+			throw new BmxException( "Failed to download the update", ex );
+		}
+
+		try {
+			string extension = Path.GetExtension( downloadUrl );
+
+			if( extension.Equals( ".zip", StringComparison.OrdinalIgnoreCase ) ) {
+				ExtractZipFile( downloadPath, currentDirectory );
+			} else if( extension.Equals( ".gz", StringComparison.OrdinalIgnoreCase ) ) {
+				ExtractTarGzipFile( downloadPath, currentDirectory );
+			} else {
+				throw new Exception( "Unknown archive type" );
```

Comment:
use BmxException with better error message

Reply:
This is already a nested exception _(not sure if we can un-nest it without making the block messy_) so I don't think the user will see it. I think the "Failed to update with new files" message should capture this tho

Reply:
ah I see. Didn't notice the nesting. I'd still like to avoid it, but it's good for now.